.circleci/config.yml

```diff
@@ -0,0 +1,86 @@
+---
+# "Include" for unit tests definition.
+unit_tests: &unit_tests
+  steps:
+    - checkout
+    - run:
+        name: Install modules and dependencies.
+        command: npm install
+    - run:
+        name: Run unit tests.
+        command: npm test
+    - run:
+        name: Submit coverage data to codecov.
+        command: npm run codecov
+        when: always
+
+version: 2.0
+workflows:
+  version: 2
+  tests:
+    jobs:
+      - node4:
+          filters:
+            tags:
+              only: /^v\d+(\.\d+){2}(-.*)?$/
+      - node6:
+          filters:
+            tags:
+              only: /^v\d+(\.\d+){2}(-.*)?$/
+      - node8:
+          filters:
+            tags:
+              only: /^v\d+(\.\d+){2}(-.*)?$/
+      - node9:
+          filters:
+            tags:
+              only: /^v\d+(\.\d+){2}(-.*)?$/
+      - publish_npm:
+          requires:
+            - node4
+            - node6
+            - node8
+            - node9
+          filters:
+            branches:
+              ignore: /.*/
+            tags:
+              only: /^v\d+(\.\d+){2}(-.*)?$/
+
+jobs:
+  node4:
+    docker:
+      - image: node:4
+        user: node
+    <<: *unit_tests
+  node6:
+    docker:
+      - image: node:6
+        user: node
+    <<: *unit_tests
+  node8:
+    docker:
+      - image: node:8
+        user: node
+    <<: *unit_tests
+  node9:
+    docker:
+      - image: node:9
+        user: node
+    <<: *unit_tests
+
+  publish_npm:
+    docker:
+      - image: node:8
+        user: node
+    steps:
+      - checkout
+      - run:
+          name: Set npm authentication.
+          command: echo "//registry.npmjs.org/:_authToken=${NPM_TOKEN}" > ~/.npmrc
+      - run:
+          name: Install modules and dependencies.
+          command: npm install
+      - run:
+          name: Publish the module to npm.
+          command: npm publish
```

Comment:
To make CI run for all commits, but publish only for v.* tags, all tasks like node4, node6, etc. should have a filter /.*/, but the publish task should be restricted to v.*.  See example here: https://github.com/googleapis/nodejs-os-login/blob/master/.circleci/config.yml - we run tests on all versions for all commits, run lint and docs for all commits, but run publish only on v* tags.

Reply:
No. The build runs for branches by default, so you don't need a filter `/.*/`. `node-h2-auto-push` has the same configuration and npm publish works fine: https://github.com/google/node-h2-auto-push/blob/master/.circleci/config.yml

Reply:
See this doc: https://github.com/google/node-h2-auto-push/blob/master/.circleci/config.yml

Reply:
Oops. Sorry. See this doc: https://circleci.com/docs/2.0/workflows/#git-tag-job-execution

Reply:
hmm. interesting
probably it's because we also have extra tasks (lint, docs, and system and sample tests for some repos) that depend on the branches?

Reply:
so what happens in our case is, {node4,node6,node7,node8,node9} -> {lint,docs} runs for all commits, but {node4,node6,node7,node8,node9} -> {lint,docs} -> publish_npm runs for tagged only, hence we probably need to have a filter of /.*/ on all tasks and a filter of /v.*/ on publish_npm.